### PR TITLE
Use new IR in function rewriter

### PR DESCRIPTION
**Pull request by titaiwangms**
Update function rewriter to use new graph IR. This avoids serialize/deserialize between rewriting. Before this PR, it's using protobuf to insert/replace functions.

Summary:
1. Instead of updating old function to use new function content, this PR uses new function to fully replace the old one in the model.
2. Avoid passing around protobuf, and modify within protobuf
3. Apply IR APIs

### CODE REVIEW

File: onnxscript/rewriter/function_rule.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 import onnxscript
 from onnxscript import ir
-from onnxscript._legacy_ir import visitor
+from onnxscript.ir import serde
```
Comment (justinchuby) [inline onnxscript/rewriter/function_rule.py:11]: Prefer using ir.serde, as a convention

**Comment by justinchuby** [comment]
Can we make the passes return an ir function? Doesnt have to be this PR.

**Comment by titaiwangms** [comment]
> Can we make the passes return an ir function? Doesnt have to be this PR.

In which we should return ir.Function?

File: onnxscript/rewriter/function_rule.py [hunk 2]
```diff
@@ -128,24 +127,14 @@ def _find_node_contains_key_in_name(
         return None
 
     def _find_node_by_type(
```
Comment (titaiwangms) [inline onnxscript/rewriter/function_rule.py:129]: @justinchuby Do we have a better approach for node searching?

File: onnxscript/rewriter/function_rule.py (hunk 1 above)
Comment (titaiwangms) [reply onnxscript/rewriter/function_rule.py]: Done

File: onnxscript/rewriter/function_rule.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/rewriter/function_rule.py:129]: Nothing better in the IR. Would it be helpful to build an auxiliary data structure here? Eg a dictionary?

Comment (titaiwangms) [reply onnxscript/rewriter/function_rule.py:129]: I think we can extend it when we have other needs.

File: onnxscript/rewriter/onnxruntime/transformers/biassplitgelu.py [hunk 3]
```diff
@@ -29,4 +29,4 @@ def ggelu(input, weight, bias):
             return msft_opset.BiasSplitGelu(matmul_input, bias)
 
         function_proto = onnxscript.script(default_opset=op)(ggelu).to_function_proto()  # type: ignore[arg-type]
```
Comment (justinchuby) [inline onnxscript/rewriter/onnxruntime/transformers/biassplitgelu.py:31]: Here for example we are returning protos. Would it make sense to return an ir function?

Comment (titaiwangms) [reply onnxscript/rewriter/onnxruntime/transformers/biassplitgelu.py:31]: I see. That makes sense. ir.Function goes in, and we should return the same type.

Comment (titaiwangms) [reply onnxscript/rewriter/onnxruntime/transformers/biassplitgelu.py:28]: Done

File: onnxscript/rewriter/function_rule.py
Comment (justinchuby) [inline onnxscript/rewriter/function_rule.py:1]: Are we able to delete _legacy_ir/ now?

File: onnxscript/rewriter/function_rule.py [hunk 4]
```diff
@@ -130,27 +127,17 @@ def _find_node_contains_key_in_name(
...
-            for output in node.output:
-                if output == value_name:
-                    return node
-        return None
-
     def compose_new_function(
-        self, old_function: onnx.FunctionProto, pkg_version: version.Version | None
-    ) -> tuple[onnx.FunctionProto, tuple[onnx.OperatorSetIdProto]]:
+        self, old_function: ir.Function, pkg_version: version.Version | None
+    ) -> tuple[ir.Function]:
```
Comment (justinchuby) [inline onnxscript/rewriter/function_rule.py:140]: Is the return type accurate? I would check if mypy is enabled for this file in lintrunner.toml

File: onnxscript/rewriter/function_rule.py
Comment (titaiwangms) [reply onnxscript/rewriter/function_rule.py:1]: optimizer/evaluator and visitor is still using legacy_ir to do graph analysis. I think that part remains unresolved. Either we decide to extend the current IR to do graph analysis or we keep the old one.

https://github.com/microsoft/onnxscript/blob/15d295a833bde1de688328be53675406eccbb82c/onnxscript/optimizer/evaluator.py#L52

@gramalingam can provide more details on this I think.

File: onnxscript/rewriter/function_rule.py (hunk 4 above)
Comment (titaiwangms) [reply onnxscript/rewriter/function_rule.py]: It looks like most/all of files under optimizer and rewriter is not covered in MYPY and PYLINT. Pattern-writing breaks MYPY.

Comment (titaiwangms) [reply onnxscript/rewriter/function_rule.py]: I enabled MYPY to cover this file.

File: onnxscript/rewriter/function_rule.py [hunk 5]
```diff
@@ -146,7 +147,8 @@ def compose_new_function(
         Raises:
             FunctionRewriteError: If the rewrite fails.
         """
-        func = self._version_controller.dispatch(pkg_version)
+        # self._version_controller is created in the subclass
```
Comment (justinchuby) [inline onnxscript/rewriter/function_rule.py:150]: hmm. That's a little strange? But we don't need to worry too much about it here

**Review by justinchuby — APPROVED** [review]

File: onnxscript/rewriter/function_rule.py [hunk 6]
```diff
@@ -185,57 +175,53 @@ def try_rewrite_function(
...
-        function.output.extend(new_function.output)
+        new_function.name = function.name
+        new_function.domain = function.domain
 
-        del function.node[:]
-        function.node.extend(nodes)
-        for new_opset in opset_imports:
-            function.opset_import.append(new_opset)
-            if new_opset.domain not in self._opset_imports:
-                model.opset_import.append(new_opset)
```
Comment (titaiwangms) [inline onnxscript/rewriter/function_rule.py:214]: This PR has a regression here. I didn't add the new opset version into opset_imports..

Comment (titaiwangms) [reply onnxscript/rewriter/function_rule.py:214]: @justinchuby Even with replace the whole new function, I think we still have to manually update opset_imports to model.opset_imports, right?

Comment (justinchuby) [reply onnxscript/rewriter/function_rule.py:214]: That's right. We could automate this process though.

Comment (justinchuby) [reply onnxscript/rewriter/function_rule.py:214]: https://github.com/microsoft/onnxscript/issues/1444

**Pull request merged**